### PR TITLE
Switch Port Modes and VLAN CLI Enhancement

### PR DESCRIPTION
**Pull request by MuhammadUmarAsad**
#### What I did

- Modified "/config/vlan.py" to add support for multiple vids, range of vids in ```vlan add|del``` and ```vlan member add|del```

- Creation of "/config/switchport.py" which will deal with port modes for interfaces i.e. access, trunk and routed

- Created the ```show interface switchport status``` and ```show interface switchport config``` commands so that it shows the status of port either being access, trunk or routed.

- Modified ```Command-Reference.md``` to add new commands in it with usage and examples

- Added new utility functions in "/utilities_common/cli.py" to use in vlan.py and switchport.py commands

- Modified test cases and added new test cases for the new commands

#### How I did it

- Added new commands and modified the existing commands in "/config/vlan.py" to 
    -  accept multiple vids 
    -  accept range of vids 
    -  ```except_flag``` and ```all option``` in vlan member command

- Creation of "/config/switchport.py" file to add commands which will deal with port modes i.e. access, trunk or routed which we were unable to define or assign in SONiC before

- New utility functions in "/utilities_common/cli.py"

- Using ```show interface switchport status``` and ```show interface switchport config``` commands, it will display if an interface is in access or trunk  or routed mode.

#### How to verify it

New commands have been added in Command-Reference.md All the syntax and examples have been added there and they can be verified by running the specific command

Signed-off-by: Muhammad Umar Asad <umarasad20@gmail.com> [ xFlow Research Inc. ]

### CODE REVIEW

File: doc/Command-Reference.md
Comment (venkatmahalingam) [inline doc/Command-Reference.md:9413]: Why cant we use the command "config vlan member add -u 1 Ethernet0" for the above?

File: doc/Command-Reference.md
Comment (venkatmahalingam) [inline doc/Command-Reference.md:9425]: If the port is not a VLAN member, by default, we assume router-port, not sure why do we need this explicit check.

File: doc/Command-Reference.md
Comment (venkatmahalingam) [inline doc/Command-Reference.md:9399]: Why cant we use the command "config vlan member add 1 Ethernet0" for the above?

File: doc/Command-Reference.md
Comment (venkatmahalingam) [inline doc/Command-Reference.md:9336]: Suggest to include range support as well with -m option instead of new range command.
For example,
sudo config vlan member add -m 100-300 Ethernet4

File: config/switchport.py
Comment (venkatmahalingam) [inline config/switchport.py:23]: Why cant we enhance the existing commands (sudo config vlan member...) to add additional validations? FYI. we do config validations in the YANG models now, please explore the same.

File: doc/Command-Reference.md
Comment (MuhammadUmarAsad) [reply doc/Command-Reference.md]: Updated as suggested

**Comment by ridahanif96** [comment]
We have modified our code along with HLD. Suggestions have been incorporated. @venkatmahalingam pls review.

**Comment by zhangyanzhao** [comment]
@venkatmahalingam can you please help to review this PR? Thanks.

File: config/vlan.py
Comment (venkatmahalingam) [inline config/vlan.py:252]: I dont think, we need 'mode' field in the PORT table.

Please check VLAN_MEMBER table to know whether the port is untagged to only one VLAN (access mode) or port is untagged member to one VLAN and tagged member to one or more VLAN, then the port mode is trunk. if the port is not present in VLAN_MEMBER table, the mode is routed.

File: scripts/intfutil
Comment (venkatmahalingam) [inline scripts/intfutil:110]: We dont need 'mode' field, we can easily infer 'access'/'trunk' for a given port using VLAN_MEMBER table.

Comment (MuhammadUmarAsad) [reply scripts/intfutil:110]: If we do not use “mode attribute” and use “VLAN_MEMBER table” instead to implement the restriction of having no tagged members in access mode then it would need to parse the VLAN_MEMBER Table which can potentially have 10’s to 1000’s of entries whenever a vlan member is added. On the other hand, by using the mode attribute we can achieve the same result using a single get command without parsing the whole VLAN_MEMBER Table.  

This mode attribute will be useful to implement other port modes as well e.g. hybrid, which is part of 202305 feature release.

Comment (venkatmahalingam) [reply scripts/intfutil:110]: I understand your point, let me think about the backward compatibility issues as well and update here.

Comment (MuhammadUmarAsad) [reply scripts/intfutil:110]: > We need the 'mode' field for portchannel as well not just for port.

So far we have focused on the physical ports only. The mode for portchannel can be added later in a separate PR.

---

> What should be the default 'mode' - 'none'?

Default mode is still “routed” as shown by the output of the “show interface status” command where it shows default behavior as routed.

---
> If we assign the IP on the port/portchannel, do we need to enforce the user to set 'routed' mode first?

In SONiC, IP assignment can be done only if the port has not assigned any VLAN. Earlier, we followed the same practice. But after your observation/concern on this, we have updated our code. We added a restriction on IP Assignment, the user needs to enforce the mode “routed” before IP assignment else it will not accept IP.

---
> Also, do we need to enforce mode config before associating the port/portchannel to VLAN(s)?

 Yes, we need to enforce mode config before associating the port to VLAN(s). For portchannel, the status is same as above answer for portchannel

---
> Do we need to restrict the mode deletion when the port is part of some VLANs?

Yes,  we need to restrict the mode deletion when a port is part of some VLAN(s)

---
> How do we handle backward compatibility? what if we load old configs after this change? are you planning to update db_migrator script?

We will get back to you about the db_migrator. Just to mention please note the following:
If we upgrade just the CLI on old versions of sonic and run commands like switch port mode, it will work and set the mode attribute in the PORT table.

---
> Please raise the SONiC YANG model pull request as well

Yes, we will raise a new PR for SONiC YANG Model.

Comment (venkatmahalingam) [reply scripts/intfutil:110]: So far we have focused on the physical ports only. The mode for portchannel can be added later in a separate PR.
>> Please add the support for portchannel in this PR itself. This helps to perform the testing for all cases and we dont have to repeat the test cases again and will be easy for the user to migrate to the validation for VLAN members.

Comment (MuhammadUmarAsad) [reply scripts/intfutil]: Please add the support for portchannel in this PR itself. This helps to perform the testing for all cases and we dont have to repeat the test cases again and will be easy for the user to migrate to the validation for VLAN members.

> > We have updated our code along with HLD as per your suggestions, pls review for PortChannel.

**Comment by zhangyanzhao** [comment]
@venkatmahalingam  can you please help to approve this PR if you are ok? Thanks.

**Comment by venkatmahalingam** [comment]
> @venkatmahalingam can you please help to approve this PR if you are ok? Thanks.

I have added @prsunny @dgsudharsan for HLD review, Once they approved the HLD, we can revisit if required and merge the PR.

**Comment by zhangyanzhao** [comment]
add @dprital  as reviewer.

**Comment by gechiang** [comment]
you should not modify the existing output of the "show interface status" command.  You should work on a new show command that displays the switchport mode.  reason is there are many applications already depended on parsing on the existing show interface status output and your change will break all those applications, parsers, and testcases...  may be create a new show command such as "show interface switch-mode".  The output mainly needs to have the switchport mode displayed in conjunction with whatever else you think may be useful for the users...

**Comment by ridahanif96** [comment]
> you should not modify the existing output of the "show interface status" command. You should work on a new show command that displays the switchport mode. reason is there are many applications already depended on parsing on the existing show interface status output and your change will break all those applications, parsers, and testcases... may be create a new show command such as "show interface switch-mode". The output mainly needs to have the switchport mode displayed in conjunction with whatever else you think may be useful for the users...

We have updated code for switchport modes configuration. Please Review @gechiang @venkatmahalingam

**Comment by ridahanif96** [comment]
@venkatmahalingam pls, review this as well.

**Comment by ridahanif96** [comment]
@dprital @qiluo-msft pls review this PR, target of this PR is 202305.

**Comment by zhangyanzhao** [comment]
@venkatmahalingam  @dprital can you please help to review and approve this PR? Thanks.

**Comment by ridahanif96** [comment]
@venkatmahalingam @dprital can you please help to review and approve this PR. Target is 202305 Thanks.

**Review by venkatmahalingam — APPROVED** [review]

**Comment by ridahanif96** [comment]
@qiluo-msft can you please merge this PR? Target is 202305 , Thanks

**Comment by ridahanif96** [comment]
@qiluo-msft can you please merge this PR, target 202305, thanks

**Comment by qiluo-msft** [comment]
@gechiang Do you have more concern?

File: config/switchport.py
Comment (gechiang) [inline config/switchport.py:125]: Should this be accepted as "pass" and not need to complain that it is already in "routed" mode?  The reason I think this is better let it pass because the interface is not yet configured with any "switchport mode" and the user perhaps just want to make sure it is set to mode "routed".  And the code actually did set the mode to "routed"...

Comment (ridahanif96) [reply config/switchport.py:125]: Yes agreed, updating this one, thanks

File: config/vlan.py
Comment (gechiang) [reply config/vlan.py:317]: Although having "switchport mode" gives more restriction control but it also takes away the liberty of freely modifying the vlan membership as needed... In my opinion, the argument could go both ways... But in my opinion it is better not to have to have the switchport mode then having it... My reason is that you should never give anyone the privilege to modify the VLAN configuration if they don't know what they are doing at the first place...  but again, this is arguable...  and for most people that came from the traditional L2 switching background and used to the concept of "Switchport" configuration this gives them the "familiar" flavor and feel of a NOS that they are used to...

**Review by gechiang — DISMISSED** [review]

**Comment by ridahanif96** [comment]
@gechiang can you please reapprove this, all tests passed. Thanks

**Review by gechiang — APPROVED** [review]

**Comment by ridahanif96** [comment]
@qiluo-msft can you please help merge this PR , thanks.

**Comment by ridahanif96** [comment]
@qiluo-msft can you please help merge this PR? thanks

**Review by theasianpianist — APPROVED** [review]
`VLAN` related changes lgtm

**Comment by ridahanif96** [comment]
@gechiang @venkatmahalingam can you please guide in this PR. Its reviewed and approved but still awaiting for merge, please help, thanks.
@vaibhavhd & @prsunny can you please help with this PR,  thanks.

**Review by gechiang — APPROVED** [review]

**Comment by gechiang** [comment]
> @gechiang Do you have more concern?

I don't have any more concerns.  Please help review and approve/merge this PR.
Thanks!

**Comment by liushilongbuaa** [comment]
@gechiang , @theasianpianist , @MuhammadUmarAsad ,
Can you please check kvmtest failure? This change is blocked by PR checker.
https://github.com/sonic-net/sonic-buildimage/pull/16267

**Comment by ridahanif96** [comment]
> @gechiang , @theasianpianist , @MuhammadUmarAsad , Can you please check kvmtest failure? This change is blocked by PR checker. [sonic-net/sonic-buildimage#16267](https://github.com/sonic-net/sonic-buildimage/pull/16267)

@liushilongbuaa I,  have checked kvmtest failure. As of my understanding this PR has not made any change that blocked/cause failure in https://github.com/sonic-net/sonic-buildimage/pull/16267 Updating sonic-utilites submodule.

**Comment by liushilongbuaa** [comment]
@ridahanif96 , it failed on vlan related command:
{"changed": false, "cmds": ["config vlan add 108", "config interface ip add Vlan108 192.168.8.1/24", "config vlan add 109", "config interface ip add Vlan109 192.168.9.1/24", "config vlan member add  108 Ethernet4", "config vlan member add  109 Ethernet4"], "delta": "0:00:02.836740", "end": "2023-08-31 16:11:34.191502", "failed": true, "msg": "At least running one of the commands failed", "results": [
{"cmd": "config vlan add 108", "cmd_with_timeout": "", "err_msg": "", "rc": 0, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": [], "timeout": 0}, 
{"cmd": "config interface ip add Vlan108 192.168.8.1/24", "cmd_with_timeout": "", "err_msg": "", "rc": 0, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": [], "timeout": 0}, 
{"cmd": "config vlan add 109", "cmd_with_timeout": "", "err_msg": "", "rc": 0, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": [], "timeout": 0}, 
{"cmd": "config interface ip add Vlan109 192.168.9.1/24", "cmd_with_timeout": "", "err_msg": "", "rc": 0, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": [], "timeout": 0}, 
{"cmd": "config vlan member add  108 Ethernet4", "cmd_with_timeout": "", "err_msg": "", "rc": 2, "stderr": "Usage: config vlan member add [OPTIONS] <vid> port\nTry \"config vlan member add -h\" for help.\n\nError: Ethernet4 is in routed mode!\nUse switchport mode command to change port mode\n", "stderr_lines": ["Usage: config vlan member add [OPTIONS] <vid> port", "Try \"config vlan member add -h\" for help.", "", "Error: Ethernet4 is in routed mode!", "Use switchport mode command to change port mode"], "stdout": "", "stdout_lines": [], "timeout": 0}, 
{"cmd": "config vlan member add  109 Ethernet4", "cmd_with_timeout": "", "err_msg": "", "rc": 2, "stderr": "Usage: config vlan member add [OPTIONS] <vid> port\nTry \"config vlan member add -h\" for help.\n\nError: Ethernet4 is in routed mode!\nUse switchport mode command to change port mode\n", "stderr_lines": ["Usage: config vlan member add [OPTIONS] <vid> port", "Try \"config vlan member add -h\" for help.", "", "Error: Ethernet4 is in routed mode!", "Use switchport mode command to change port mode"], "stdout": "", "stdout_lines": [], "timeout": 0}
], "start": "2023-08-31 16:11:31.354762"}

**Comment by ridahanif96** [comment]
All vlan related failure issues resolved.

File: config/vlan.py
Comment (qiluo-msft) [inline config/vlan.py:325]: >port_data [](http://example.com/codeflow?start=29&length=9)

PORT/mode is not yang modelled. So this PR will become breaking change, and will block this submodule move. I will raise a PR to revert this PR.

Comment (ridahanif96) [reply config/vlan.py:325]: @qiluo-msft , 
@venkatmahalingam @gechiang We have fixed YANG Model for this PR. The [YANG PR](https://github.com/sonic-net/sonic-buildimage/pull/13580) is failing as it needs supported utilties sourcecode. One posssible solution is that we merge https://github.com/sonic-net/sonic-utilities/pull/2419 utilities code and then we merge this PR. All dependencies will be be solved. 
Other solution is that we remove MUST condition from this PR for instance in sonic-vlan.yang for Vlan membership for only on access and trunk port and merge this PR and merge utilities source code. After that i will add a new PR with must condition check on sonic-vlan membership.It is requested to please help in this matter. Thanks!

**Pull request merged**
